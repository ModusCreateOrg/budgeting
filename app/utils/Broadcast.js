// @flow

type Subscription = (currentState: mixed) => void;

/**
 * Broadcast
 *
 * Event emitter which provides a generic way to pass values to children
 * using React's context.
 *
 * We have to use this because we can't send data to children directly in the context.
 * If we do so, and `shouldComponentUpdate` returns false somewhere in the child tree,
 * then it's children won't get the updated context when it changes.
 *
 * To be able to send data, context should not change. Instead, context is used
 * to notify components about state changes.
 */
class Broadcast {
  currentState: mixed;
  subscriptions: Subscription[];

  constructor(initialState: mixed): void {
    this.currentState = initialState;
    this.subscriptions = [];
  }

  getState = (): mixed => this.currentState;

  setState = (newState: mixed): void => {
    this.currentState = newState;

    this.subscriptions.forEach(subscription => subscription(this.currentState));
  };

  subscribe = (subscription: Subscription): (() => void) => {
    this.subscriptions.push(subscription);

    // return `unsubscribe` function
    return () => {
      this.subscriptions = this.subscriptions.filter(item => item !== subscription);
    };
  };
}

export default Broadcast;
