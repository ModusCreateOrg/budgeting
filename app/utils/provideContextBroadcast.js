// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import getDisplayName from 'utils/getDisplayName';
import Broadcast from 'utils/Broadcast';
import type { HigherOrderComponent } from 'types';

type ProvidedProps = {
  setBroadcastState: (newState: mixed) => void,
};

/**
 * Returns a HOC you can use to send a Broadcast object through context.
 * It passes a `setBroadcastState` prop to the wrapped component to update
 * the Broadcast state.
 */
const provideContextBroadcast = (
  broadcastName: string
): HigherOrderComponent<{}, ProvidedProps> => WrappedComponent => {
  class ContextBroadcastProvider extends React.Component<{}> {
    static childContextTypes = {
      [broadcastName]: PropTypes.object,
    };

    constructor(props, context) {
      super(props, context);

      // Initialize form data broadcast
      this.broadcast = new Broadcast();
    }

    getChildContext() {
      return { [broadcastName]: this.broadcast };
    }

    broadcast: Broadcast;

    render() {
      return <WrappedComponent setBroadcastState={this.broadcast.setState} {...this.props} />;
    }
  }

  ContextBroadcastProvider.displayName = `ContextBroadcastProvider(${getDisplayName(WrappedComponent)})`;
  return ContextBroadcastProvider;
};

export default provideContextBroadcast;
