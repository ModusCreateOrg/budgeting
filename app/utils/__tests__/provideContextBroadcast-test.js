import * as React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import Broadcast from '../Broadcast';
import provideContextBroadcast from '../provideContextBroadcast';

describe('provideContextBroadcast', () => {
  it('should return a HOC that renders correctly', () => {
    const hoc = provideContextBroadcast('testBroadcast');
    const WrappedComponent = () => <div>test</div>;
    const GeneratedComponent = hoc(WrappedComponent);

    const tree = renderer.create(<GeneratedComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return a HOC that sets a Broadcast object in the context', () => {
    const hoc = provideContextBroadcast('testBroadcast');

    class WrappedComponent extends React.Component {
      static contextTypes = {
        testBroadcast: PropTypes.object,
      };

      render() {
        const broadcast = this.context.testBroadcast;
        expect(broadcast instanceof Broadcast).toEqual(true);

        return <div>test</div>;
      }
    }

    const GeneratedComponent = hoc(WrappedComponent);
    renderer.create(<GeneratedComponent />).toJSON();
  });

  it('should return a HOC with a prop to update the state of the Broadcast object in the context', () => {
    const hoc = provideContextBroadcast('testBroadcast');

    // eslint-disable-next-line react/no-multi-comp
    class WrappedComponent extends React.Component {
      static contextTypes = {
        testBroadcast: PropTypes.object,
      };

      componentWillMount() {
        const { setBroadcastState } = this.props;

        // update broadcast state
        setBroadcastState('newState');
      }

      render() {
        const broadcast = this.context.testBroadcast;

        // check that the broadcast state was updated
        //
        // NOTE: In real code this should be checked using a subscriber, that handles the case of
        // `shouldComponentUpdate` returning `false` in any place between the context provider and
        // the component that subscribes.
        expect(broadcast.getState()).toEqual('newState');

        return <div>test</div>;
      }
    }

    const GeneratedComponent = hoc(WrappedComponent);
    renderer.create(<GeneratedComponent />).toJSON();
  });
});
