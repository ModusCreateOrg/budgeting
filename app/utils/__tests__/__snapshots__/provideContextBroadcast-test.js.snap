// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provideContextBroadcast should return a HOC that renders correctly 1`] = `
<div>
  test
</div>
`;
